src/systems/filecoin_vm/runtime/runtime.go

```diff
@@ -1,81 +1,424 @@
+	rt._checkRunning()
+	rt._checkStateLock(false)
+
+	transferRet := rt._globalStatePending().WithFundsTransfer(from, to, amount)
+	if transferRet.Which() == st.StateTree_WithFundsTransfer_FunRet_Case_error {
+		rt._throwError(exitcode.SystemError(exitcode.InsufficientFunds))
+	} else {
+		rt._globalStatePending_ = transferRet.As_StateTree()
+	}
+}
+
+func (rt *Runtime_I) _incrementCallSeqNum(a addr.Address) {
 	panic("TODO")
 }
 
-func (s *VMState_I) ComputeActorAddress(creator addr.Address, nonce actor.CallSeqNum) addr.Address {
+// TODO: This function should be private (not intended to be exposed to actors).
+// (merging runtime and interpreter packages should solve this)
+func (rt *Runtime_I) SendToplevelFromInterpreter(to addr.Address, input InvocInput, ignoreErrors bool) (
+	msg.MessageReceipt, st.StateTree) {
+
+	rt._running_ = true
+	ret := rt._sendInternal(to, input, ignoreErrors)
+	rt._running_ = false
+	return ret, rt._globalStatePending()
+}
+
+func _invokeMethodInternal(
+	rt *Runtime_I,
+	actorCode ActorCode,
+	method actor.MethodNum,
+	params actor.MethodParams) (ret InvocOutput, gasUsed msg.GasAmount) {
+
+	if method == Reserved_NoopMethod {
+		ret = msg.InvocOutput_Make(exitcode.OK(), nil)
+		gasUsed = msg.GasAmount_Zero() // TODO: verify
+		return
+	}
+
+	rt._running_ = true
+	catchRetCode := exitcode.CatchRuntimeErrors(func() {
+		actorCode.InvokeMethod(rt, method, params)
+		rt._checkStateLock(false)
+		rt._checkNumValidateCalls(1)
+		rt._checkNumReturnCalls(1)
+		for _, item := range rt._messageQueue() {
+			rt._sendInternal(item.to, item.input, item.ignoreErrors)
+		}
+		rt._messageQueue_ = []MessageQueueItem{}
+	})
+	rt._running_ = false
+
+	if catchRetCode.IsSuccess() {
+		if rt._output() == nil {
+			panic("Internal error: return call recorded but no output object returned")
+		}
+		ret = rt._output()
+	} else {
+		ret = msg.InvocOutput_Make(catchRetCode, nil)
+	}
+
+	// TODO: Update gasUsed
+	TODO()
+
+	return
+}
+
+func (rtOuter *Runtime_I) _sendInternal(to addr.Address, input InvocInput, ignoreErrors bool) msg.MessageReceipt {
+	rtOuter._checkRunning()
+	rtOuter._checkStateLock(false)
+
+	toActor := rtOuter._globalStatePending().GetActor(to).State()
+
+	toActorCode, err := loadActorCode(toActor.CodeCID())
+	if err != nil {
+		rtOuter._throwError(exitcode.SystemError(exitcode.ActorCodeNotFound))
+	}
+
+	var toActorMethodGasBound msg.GasAmount
+	TODO() // TODO: obtain from actor registry
+	rtOuter._deductGasRemaining(toActorMethodGasBound)
+	// TODO: gasUsed may be larger than toActorMethodGasBound if toActor itself makes sub-calls.
+	// To prevent this, we would need to calculate the gas bounds recursively.
+
+	rtOuter._transferFunds(rtOuter._actorAddress(), to, input.Value())
+	rtOuter._incrementCallSeqNum(to)
+
+	rtInner := Runtime_Make(
+		rtOuter._globalStatePending(),
+		to,
+		input.Value(),
+		rtOuter._gasRemaining(),
+	)
```

Comment:
- why are we making another VM runtime -- the intention is for the VM to only have one vm.Runtime.
- The runtime is meant to be a single syscalls-like layer, hopefully holding minimal state (or simply proxying calls to sub objects)
- I believe you're now using the runtime as the invocations pair (`InvocInput/InvocOutput`) were meant to be used. If you don't want the pair, maybe use a third object `Invocation`, or `ActorRuntime` to hold the invocation state changes? (i understand the intention of having each actor be isolated/sandboxed -- so  maybe `ActorRuntime`? `InvocationRuntime`? `InvocRuntime`? `MethodRuntime` ?

Reply:
(moved state into separate VMContext struct as per discussion)